### PR TITLE
A leftmost rule is applied on staves mapping, and a rightmost rule is applied on system mapping

### PR DESCRIPTION
**Pull request by cdecron**
Lots of validation examples are different from previous version but it's not relevant (tiny mapping changes).

This new behavior still doesn't solve all cases.

`{ [ a | b ], [ c | _/8 ] }`
image
This example fails because `_/8` is shorter than `b`, which implies that `mTimeAndBoxRightmostCompare` sorts `_/8` before, whereas `b` should be higher in the list. Handling this in `reduce` is doable but produces several other mapping problems.
This problem appears on maps/system/regression-tests/Ibert-First10Measures.svg, system 2, measure 2.
image

`{ [ \meter<"C"> c/1 _  ],
  [ \staff<1> \meter<"C"> empty/1 _ ] }`
image
This example fails too because there are two rests at the same date on the same staff (voice 1 an 2).

To finish, maps/system/gmn-examples/complex_files/846_2f.svg produces a bad artifact of this new mechanism.

### CODE REVIEW

File: src/engine/lib/GUIDOScoreMap.cpp [hunk 1]
```diff
@@ -86,6 +86,12 @@ bool TimeSegment::include(const TimeSegment& ts) const
 	return (s1a <= s2a) && (s1b >= s2b);
 }
 
+bool TimeSegment::startEqual(const TimeSegment& ts) const
```
Comment (dfober) [inline src/engine/lib/GUIDOScoreMap.cpp]: you could miss dates that are equal but expressed using different fractions (e.g. 2/4 and 1/2). I suggest to use Fraction or to compare using a float reduction.

File: src/engine/maps/GuidoMapCollector.cpp [hunk 2]
```diff
@@ -117,6 +117,7 @@ void GuidoVoiceAndBarCollector::Graph2TimeMap( const FloatRect& box, const TimeS
 {
     bool isBar = (infos.type == kBar || infos.type == kRepeatBegin || infos.type == kRepeatEnd);
     
+    if ( fStaffNum != 0 && fStaffNum != infos.staffNum)   return; // other staves are filtered out if fStaffNum is defined
```
Comment (dfober) [inline src/engine/maps/GuidoMapCollector.cpp]: you should use parenthesis for the second part of the test:
if ( fStaffNum != 0 && fStaffNum != infos.staffNum)
it's not for safety but for readability (I know the precedence rule will apply)

**Comment by dfober** [comment]
you should find some inline comments regarding the changes.
There is a pending conflict that prevents me from a rebase and merge. Could you solve it? (maybe updating your local repo)

**Comment by cdecron** [comment]
I made appropriate corrections.
Actually I see that "This branch has no conflicts with the base branch", so I guess it should be ok to merge now?

**Comment by dfober** [comment]
>

>  Actually I see that "This branch has no conflicts with the base branch", so I guess it should be ok to merge now?

Right for a "merge commit" but not for a "rebase and merge". Try a rebase and merge pull.

**Pull request closed without merging**